### PR TITLE
device listing & updates

### PR DESCRIPTION
**Pull request by bboozzoo**
Fixes MEN-992

Also a bunch of updates:
* replaced deviceauth versions of requestlog, accesslog, requestid, log with go-lib-micro equivalents
* updated tests to use `github.com/stretchr/testify/mock`
* replaced hand crafted mocks with ones generated with mockery

@mendersoftware/rndity @maciejmrowiec

### CODE REVIEW

File: docs/management_api.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          schema:
+            $ref: '#/definitions/Error'
+definitions:
+  Device:
+    type: object
+    properties:
+      id:
+        type: string
+        description: Mender assigned Device ID.
+      tenant_token:
```
Comment (maciejmrowiec) [inline docs/management_api.yml]: @mchalski should tenant_token be here? All devices would have same within request.

File: datastore_mongo.go [hunk 2]
```diff
@@ -79,6 +79,21 @@ func NewDataStoreMongo(host string) (*DataStoreMongo, error) {
...
 
+func (db *DataStoreMongo) GetDevices(skip, limit uint) ([]Device, error) {
+	s := db.session.Copy()
+	defer s.Close()
+
+	c := s.DB(DbName).C(DbDevicesColl)
+
+	res := []Device{}
+
+	err := c.Find(bson.M{}).Skip(int(skip)).Limit(int(limit)).All(&res)
```
Comment (maciejmrowiec) [inline datastore_mongo.go]: Need to sort on _id before skip/limit otherwise several requests for the same page will end up with different results.

**Comment by bboozzoo** [comment]
btw. still need to add some tests for `/devices` and `/devices/{id}` endpoints

File: api_devauth.go [hunk 3]
```diff
@@ -141,6 +148,51 @@ func (d *DevAuthHandler) SubmitAuthRequestHandler(w rest.ResponseWriter, r *rest
...
+	}
+
+	l.Infof("page: %v perpage: %v", page, perPage)
+	skip := (page - 1) * perPage
+	limit := perPage + 1
+	if devs, err := da.GetDevices(uint(skip), uint(limit)); err != nil {
+		restErrWithLogInternal(w, r, l, err)
+	} else {
+		w.WriteJson(devs)
+	}
```
Comment (kjaskiewiczz) [inline api_devauth.go]: should we add page link headers to the response?

File: datastore_mongo.go (hunk 2 above)
Comment (bboozzoo) [inline datastore_mongo.go]: fixed

File: api_devauth.go (hunk 3 above)
Comment (bboozzoo) [inline api_devauth.go]: fixed

**Comment by bboozzoo** [comment]
Pushed a couple of patches adding acceptance tests.

File: api_devauth.go [hunk 4]
```diff
@@ -141,6 +148,65 @@ func (d *DevAuthHandler) SubmitAuthRequestHandler(w rest.ResponseWriter, r *rest
...
+		restErrWithLogInternal(w, r, l, err)
+	}
+
+	page, perPage, err := rest_utils.ParsePagination(r)
+	if err != nil {
+		restErrWithLog(w, r, l, err, http.StatusBadRequest)
+		return
+	}
+
+	l.Infof("page: %v perpage: %v", page, perPage)
```
Comment (kjaskiewiczz) [inline api_devauth.go]: unnecessary info log

Comment (bboozzoo) [inline api_devauth.go]: fixed

**Comment by bboozzoo** [comment]
If there are no more comments, I'll be merging this later today.

File: api_devauth_test.go [hunk 5]
```diff
@@ -446,16 +442,151 @@ func TestApiDevAuthDeleteToken(t *testing.T) {
...
+			skip:  2,
+			limit: 3,
+			code:  http.StatusInternalServerError,
+			err:   errors.New("failed"),
+			body:  RestError("internal error"),
+		},
+	}
+
+	for i, tc := range tcases {
+		t.Logf("tc %v", i)
```
Comment (maciejmrowiec) [inline api_devauth_test.go]: there is t.Run (subtest) functionality that can be used for logging/execution of test cases, for sure slightly cleaner and can run test cases in parallel, execute selected subtest with -run. Usually trying to switch into that when I see log used but this is matter of taste at this point.

https://blog.golang.org/subtests

Comment (bboozzoo) [inline api_devauth_test.go]: I'll check it out. If it's usable I'll file a ticket with improvement on Jira as we'll probably want to adopt it everywhere.

Comment (bboozzoo) [inline api_devauth_test.go]: Actually turned out to be quite easy, so I've updated the tests.

**Comment by bboozzoo** [comment]
Pushed a couple more commits that introduce `t.Run()` subtests, and add `t.Parallel()` where possible (basically cuts test execution time in half on my machine)

**Comment by bboozzoo** [comment]
Ok, so the same set of tests, but running in parallel yields lower coverage on codecov? seriously?

**Comment by maciejmrowiec** [comment]
thats kind of ... weird
did you to run locally and compare coverage report spit by go test?

**Comment by bboozzoo** [comment]
Updated `global: unit tests: mark tests as parallel where possible` with proper fix.

**Comment by bboozzoo** [comment]
merge?

**Pull request merged**